### PR TITLE
Handle harvester loading in a route guard to better handle the unauthenticated method.

### LINKED ISSUE #11350: [BUG] Refresh browser for the virtualization management page of accessing Harvester would display empty or page not found error
**Rancher Server Setup**
- Rancher version: `v2.9.0-alpha5` and `v2.9-head` (4c5e874)
- Installation option (Docker install/Helm Chart): Helm Chart

**Harvester Setup**
- Harvester version: `v1.3.1` and `v1.2.2`

**Describe the bug**

With Harvester have imported in Rancher
When access Harvester in Rancher virtualization management. 

* After the Harvester cluster content come up
    image

* On Harvester **v1.3.1** + Rancher **v2.9.5-alpha5**
  - When we try to fresh the page via browser function (or F5), It would display empty page content 
  
    https://github.com/rancher/dashboard/assets/29251855/46dee9f4-3a7a-4578-9091-2134e90798a0

* On Harvester **v1.2.2** + Rancher **v2.9-head** 
  - When we try to fresh the page via browser function (or F5), It would display `Error 404: This page could not be found`
  
  image

**To Reproduce**

1. Use helm to provision the Rancher v2.9.0-alpha5 
    - Repo url: https://releases.rancher.com/server-charts/alpha
2. Create a v1.3.1 Harvester cluster
3. Import Harvester to Rancher
4. Access Harvester from Rancher virtualization management page
5. Refresh the page

**Expected Result**

Every time refresh the Harvester cluster content page should display all content well.

### PR DESCRIPTION
**Pull request by codyrancher**
### Summary
This is done by providing a 404l page which is authenticated.

Fixes #11350

### Technical notes summary
We will still need the fix from https://github.com/rancher/dashboard/pull/11359/commits/be1e27b0630d32d2309c5d3969643f5fa84df014 for prod.

### Areas or cases that should be tested
404 routing.
Harvester routing.

### Screenshot/Video
Regular 404:
https://github.com/rancher/dashboard/assets/55104481/5d3d2fc8-0cbd-4d20-99e0-59dda8a0d8e3

Harvester 404:
https://github.com/rancher/dashboard/assets/55104481/310f10e4-6881-402f-8952-a55f9b846d82

I noticed a flicker of the resource page on this one. I can only reproduce this on the harvester cluster  but it's in `master` and unrelated to both this PR and harvester itself. If anyone else can reproduce this I'll resolve it as a separate issue.

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone 
- [x] The PR template has been filled out
- [x] The PR has been self reviewed 
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

### CODE REVIEW

File: shell/config/router/navigation-guards/harvester.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import dynamicPluginLoader from '@shell/pkg/dynamic-plugin-loader';
+import { routeRequiresAuthentication } from '@shell/utils/router';
+
+export function install(router, context) {
+  router.beforeEach((to, from, next) => loadHarvester(to, from, next, context));
+}
+
+export async function loadHarvester(to, from, next, { store }) {
+  if (!routeRequiresAuthentication(to) || to.name !== '404') {
```
Comment (codyrancher) [inline shell/config/router/navigation-guards/harvester.js:9]: Only execute if authenticated and on a 404 page.

File: shell/config/router/routes.js [hunk 2]
```diff
@@ -484,4 +484,12 @@ export default [
         component: () => interopDefault(import('@shell/pages/c/_cluster/_product/_resource/_namespace/_id.vue')),
         name:      'c-cluster-product-resource-namespace-id'
       }]
-  }];
+  },
+  {
+    path:      '*',
+    name:      '404',
+    component: () => interopDefault(import('@shell/pages/404.vue')),
+    meta:      { requiresAuthentication: true },
```
Comment (codyrancher) [inline shell/config/router/routes.js:492]: We specifically needed this component so we could make sure authentication would be run on 404s.

File: shell/pages/404.vue [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<script>
+import Brand from '@shell/mixins/brand';
+
+export default {
+  mixins: [Brand],
```
Comment (codyrancher) [inline shell/pages/404.vue:5]: We need this to reduce the flicker in dark mode.

File: shell/pages/404.vue [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  beforeMount() {
+    this.$store.commit('setError', { error: new Error('404: This page could not be found') });
+    this.$router.replace('/fail-whale');
+  }
```
Comment (codyrancher) [inline shell/pages/404.vue:9]: If route guards don't redirect us we should go to failwhale with the error message we used to emit in the `render()` method of `entry-helpers.js`

File: shell/initialize/entry-helpers.js [hunk 5]
```diff
@@ -195,26 +194,6 @@ async function render(to, from, next) {
   const matches = [];
   const Components = getMatchedComponents(to, matches);
 
-  // If no Components matched, generate 404
-  if (!Components.length) {
```
Comment (codyrancher) [inline shell/initialize/entry-helpers.js:199]: This will no longer ever be true because our 404 page will match if nothing else will.

**Comment by codyrancher** [comment]
I'll open for review once tests pass again. Looks like we may have had a node version change. This change was passing all unit and e2e tests on friday.

File: shell/config/router/navigation-guards/runtime-extension-route.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+export async function runtimeExtensionRoute(to, from, next, { store }) {
+  if (!routeRequiresAuthentication(to) || to.name !== '404') {
+    return next();
+  }
+
+  try {
+    // Handle the loading of dynamic plugins (Harvester) because we only want to attempt to load those plugins and routes if we first couldn't find a page.
+    // We should probably get rid of this concept entirely and just load plugins at the start.
+    await store.dispatch('loadManagement');
```
Comment (richard-cox) [inline shell/config/router/navigation-guards/runtime-extension-route.js:16]: This makes me nervous, we're calling this before some of the normal setup stuff in the auth middleware. It's quite dependent on the order the nav guards run as well, more in PR review comment

**Review by richard-cox — APPROVED** [review]
When testing I saw all nav guards start in order, but finish out of order. This makes me very nervous, but i couldn't find any specific bugs. We should pay attention to this, in particular things like multiple api requests to the same endpoint on load, or extra schema missing warnings. Just as a tidy up I added async/await to the beforeEach hooks (though i don't think vue router `awaits` them). 

I also de-harvester'd the navigation guard names

**Comment by codyrancher** [comment]
I think we need to be really clear what the issue is because the way that you've described it and the commit you added makes me think there's a misunderstanding of how navigation guards work.

> Just as a tidy up I added async/await to the beforeEach hooks (though i don't think vue router awaits them).

You're right, vue-router isn't awaiting them. It's waiting for the the hook passed to `beforeEach` to call `next()`. This change doesn't actually safeguard against anything.

> When testing I saw all nav guards start in order, but finish out of order.

The call to `next()` is what will call the next navigation guard in the list. You will not see the caller finish until the callee finishes because it's going in stack order.

```
function b() {
   console.log('b');
}

function a() {
   console.log('a')
   b()
   console.log('c')
}

a();
```

We will see `abc` printed which indicates `a()` finished executing after `b()` finished. This is the exact behavior we should see with the beforeEach navigation guards. If you're seeing out of stack order then we do have a problem and we should address that. But that's not what I noticed while working on this.

**Comment by codyrancher** [comment]
I'll go ahead and merge once master has all tests pass or we decide to skip.

**Pull request merged**